Integration issues with react-native-gesture-handler on Android

I am attempting to use `Navigation React Native` as the routing solution for my native app and am liking it quite a bit: performant, simple, intuitive and quite flexible. However, In my early integration attempts (using some of the other core libraries i would be using in my redux-driven app) i find that I have a couple of Android-specific issues when i attempt integration with [`react-native-gesture-handler` ](https://github.com/kmagiera/react-native-gesture-handler). 

### Main Issue
The [`SideDrawer`](https://kmagiera.github.io/react-native-gesture-handler/docs/component-drawer-layout.html) seems to work only in the initial screen  when an attempt is made to trigger it via a gesture (swipe left/right from the edges, depending on the configuration). When i navigate away to another screen that also has the SideDrawer, it does not work. It does work everywhere when i try to programmatically open/close, regardless of which screen i am on, however. 

### Secondary Issue
Additionally a minor issue was noticed as well. Running back actions, both via an app button and via the hardware back button, seem to flash the side drawer for about a second before navigation. It seems like it is navigating 2 steps instead of 1 with the flashing drawer seeming to visually appear as an intermediary screen. 

_**Note:** Everything works perfectly fine in IoS however._

I use the SideDrawer as an HOC, wrapping screens that would require the SideDrawer but have attempted using it normally as well.

The [installation docs](https://kmagiera.github.io/react-native-gesture-handler/docs/getting-started.html) for `react-native-gesture-handler` has a [section](https://kmagiera.github.io/react-native-gesture-handler/docs/getting-started.html#with-wix-react-native-navigation-https-githubcom-wix-react-native-navigation) dedicated to installation for android apps that use native navigation libraries. Essentially they require wrapping each screen with a dedicated HOC shipped with the library. 

As far i understand `Navigation` has a single ReactRoot([a recent change](https://github.com/grahammendick/navigation/pull/245)) and so, if wrapping is necessary, it can be done at the topmost level. I tried that but It didn't work. I also attempted wrapping individual screens but with no success. Ive used the gesture handler library in other apps, both JS-driven([`react-navigation`](https://github.com/react-navigation/react-navigation), no change was necessary) and native ([react-native-navigation](https://github.com/wix/react-native-navigation), with integration of the specific changes required as was documented) and they work fine. My guess is i might perhaps be needing the wrapping of the dedicated HOC shipped with react-native-gesture-handler since, as per the [installation docs](https://grahammendick.github.io/navigation/documentation/native/setup.html) for `Navigation` on React Native, there is specific mention of the use of native APIs for navigation:

> The Navigation router for React Native uses the underlying native APIs to provide faithful navigation on Android and iOS. 




My current minimal package.json:

```
{
  "name": "menuappnative",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "start": "node node_modules/react-native/local-cli/cli.js start",
    "test": "jest"
  },
  "dependencies": {
    "@react-native-community/async-storage": "^1.5.1",
    "@react-native-community/netinfo": "^4.0.0",
    "navigation": "^5.2.0",
    "navigation-react": "^4.1.1",
    "navigation-react-native": "^5.3.0",
    "prop-types": "^15.7.2",
    "react": "16.8.3",
    "react-native": "0.59.10",
    "react-native-device-info": "^2.2.2",
    "react-native-firebase": "^5.5.5",
    "react-native-gesture-handler": "^1.3.0",
    "react-native-vector-icons": "^6.6.0",
    "react-redux": "^7.1.0",
    "redux": "^4.0.3",
    "redux-thunk": "^2.3.0"
  },
  "devDependencies": {
    "@babel/core": "^7.5.4",
    "@babel/runtime": "^7.5.4",
    "babel-jest": "^24.8.0",
    "jest": "^24.8.0",
    "metro-react-native-babel-preset": "^0.55.0",
    "react-test-renderer": "16.8.3"
  },
  "jest": {
    "preset": "react-native"
  }
}
```

The App.js:

```
import React from 'react';
import { StateNavigator } from 'navigation';
import { NavigationHandler } from 'navigation-react';
import { NavigationStack } from 'navigation-react-native';
import { Alert, BackHandler } from 'react-native';
import { Provider } from 'react-redux';
import Menu from './Menu';
import store from './src/redux/store';
import Welcome from './Welcome';

const stateNavigator = new StateNavigator([
  {key: 'welcome', trackCrumbTrail: true},
  {key: 'menu', trackCrumbTrail: true},
]);

const {welcome,menu} = stateNavigator.states;
welcome.renderScene = () => <Welcome />;
menu.renderScene = () => <Menu />;
stateNavigator.navigate('welcome');

class App extends React.Component{
  componentDidMount(){
    BackHandler.addEventListener("hardwareBackPress",this._handleHardwareBackPress)
  }
  componentWillUnmount(){
    BackHandler.removeEventListener("hardwareBackPress",this._handleHardwareBackPress)
  }
  _handleHardwareBackPress=(args)=>{
      if(!stateNavigator.canNavigateBack(1)){
        Alert.alert("Show a prompt with OK/CANCEL")
        return true
      }
  }
  render(){
    return(
      <Provider store={store}>
          <NavigationHandler stateNavigator={stateNavigator}>
            <NavigationStack />
          </NavigationHandler>
      </Provider>
    );
  }
}
export default App;
```

The Welcome screen:

```
import React, { Component } from 'react';
import { NavigationContext } from 'navigation-react';
import { Button, SafeAreaView, StyleSheet, Text } from 'react-native';
import { connect } from 'react-redux';
//import SideDrawer from './src/modules/SideDrawer/SideDrawer';
import withSideDrawer from './src/helpers/hoc/withSideDrawer';

class Welcome extends Component {
  _onPress = (stateNavigator) => {
    stateNavigator.navigate('menu');
  }
  render() {
    return (
      <NavigationContext.Consumer>
        {({stateNavigator}) => (
          <SafeAreaView style={styles.container}>
            <Text>
              Welcome
            </Text>
              <Button title="To Menu Screen" onPress={() => this._onPress(stateNavigator)}/>
          </SafeAreaView>
        )}
      </NavigationContext.Consumer>
    );
  }
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#E91E63'
  },
});

const mapStateToProps = (state) => {
  return {common: state.commonReducer, 
    nav: state.navigationReducer}
}
export default connect(mapStateToProps)(withSideDrawer(Welcome));
```

The Menu Screen

```
import { NavigationContext } from 'navigation-react';
import React, { Component } from 'react';
import { Button, SafeAreaView, StyleSheet, Text } from 'react-native';
import { connect } from 'react-redux';
import withSideDrawer from './src/helpers/hoc/withSideDrawer';

class Menu extends Component {
  _onPress = (stateNavigator) => {
    stateNavigator.navigateBack(1);
  }
  render() {
    return (
      <NavigationContext.Consumer>
        {({stateNavigator}) => (
          <SafeAreaView style={styles.container}>
            <Text>
              Menu
            </Text>
              <Button title="Back" onPress={() => this._onPress(stateNavigator)}/>
          </SafeAreaView>
        )}
      </NavigationContext.Consumer>
    );
  }
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#E91E63'
  },
});

const mapStateToProps = (state) => {
  return {common: state.commonReducer, nav: state.navigationReducer}
}
export default connect(mapStateToProps)(withSideDrawer(Menu));
```

I had followed the normal(non-native navigation libraries) installation procedure already so didn't have to change anything in `MainApplication.java`. In other words all my attempts prior to raising the issue were made with this already in place.

These were my 2 broad attempts and outcomes.

### Approach 1:
- Rollback the additions in `MainApplication.java` and treat as a native installation and follow the RNN section. [This is just to rule-out this approach if it doesn't work]
- Add the HOC in my files as suggested. In App.js or individually in each screen that will be rendered

**Outcome:** Crashes without opening when i attempt navigating to the inner screen i.e Welcome opens but cannot navigate to Menu from the button link. The app crashes. I don't have my own HOC and instead call the `DrawerLayout` component directly inline to simplify things.


### Approach 2
- Follow the normal non-native installation procedure outlined in the `react-native-gesture-handler` documentation
- Follow your suggestions above about addition to `SceneActivity.java`
- Try with the HOC
- Try without the HOC

**Outcome:** My not upto-the-mark(**an understatement**) java skills has me stuck at integrating the changes you had outlined. In the `node_modules` directory i navigated to the `SceneActivity.java` file and made changes so that the file now reads:

```
package com.navigation.reactnative;

import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

import com.facebook.react.ReactActivity;
import com.facebook.react.bridge.GuardedRunnable;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler;
import com.facebook.react.uimanager.JSTouchDispatcher;
import com.facebook.react.uimanager.RootView;
import com.facebook.react.uimanager.UIManagerModule;
import com.facebook.react.uimanager.events.EventDispatcher;
import com.facebook.react.views.view.ReactViewGroup;


 import com.facebook.react.ReactActivityDelegate;
 import com.facebook.react.ReactRootView;
 import com.swmansion.gesturehandler.react.RNGestureHandlerEnabledRootView;


import java.util.HashSet;

public class SceneActivity extends ReactActivity implements DefaultHardwareBackBtnHandler {
    public static final String CRUMB = "Navigation.CRUMB";
    public static final String SHARED_ELEMENTS = "Navigation.SHARED_ELEMENTS";
    private SceneRootViewGroup rootView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        int crumb = getIntent().getIntExtra(CRUMB, 0);
        rootView = new SceneRootViewGroup(getReactNativeHost().getReactInstanceManager().getCurrentReactContext());
        if (crumb < NavigationStackView.sceneItems.size()) {
            View view = NavigationStackView.sceneItems.get(crumb).view;
            if (view.getParent() != null)
                ((ViewGroup) view.getParent()).removeView(view);
            rootView.addView(view);
        }
        setContentView(rootView);
        @SuppressWarnings("unchecked")
        HashSet<String> sharedElements = (HashSet<String>) getIntent().getSerializableExtra(SHARED_ELEMENTS);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && sharedElements != null ) {
            this.postponeEnterTransition();
            SharedElementTransitioner transitioner = new SharedElementTransitioner(this, sharedElements);
            findViewById(android.R.id.content).getRootView().setTag(R.id.sharedElementTransitioner, transitioner);
        }
    }

    @Override
    public void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        int crumb = intent.getIntExtra(CRUMB, 0);
        if (rootView.getChildCount() > 0)
            rootView.removeViewAt(0);
        if (crumb < NavigationStackView.sceneItems.size()) {
            View view = NavigationStackView.sceneItems.get(crumb).view;
            if (view.getParent() != null)
                ((ViewGroup) view.getParent()).removeView(view);
            rootView.addView(view);
        }
    }
    @Override
    protected String getMainComponentName() {
        return "navigation";
    }

     @Override
      protected ReactActivityDelegate createReactActivityDelegate() {
        return new ReactActivityDelegate(this, getMainComponentName()) {
          @Override
          protected ReactRootView createRootView() {
           return new RNGestureHandlerEnabledRootView(SceneActivity.this);
          }
        };
      }


    static class SceneRootViewGroup extends ReactViewGroup implements RootView {
        private boolean hasAdjustedSize = false;
        private int viewWidth;
        private int viewHeight;

        private final JSTouchDispatcher mJSTouchDispatcher = new JSTouchDispatcher(this);

        public SceneRootViewGroup(Context context) {
            super(context);
        }

        @Override
        protected void onSizeChanged(final int w, final int h, int oldw, int oldh) {
            super.onSizeChanged(w, h, oldw, oldh);
            viewWidth = w;
            viewHeight = h;
            updateFirstChildView();
        }

        private void updateFirstChildView() {
            if (getChildCount() > 0) {
                hasAdjustedSize = false;
                final int viewTag = getChildAt(0).getId();
                ReactContext reactContext = getReactContext();
                reactContext.runOnNativeModulesQueueThread(
                    new GuardedRunnable(reactContext) {
                        @Override
                        public void runGuarded() {
                            (getReactContext()).getNativeModule(UIManagerModule.class)
                                .updateNodeSize(viewTag, viewWidth, viewHeight);
                        }
                    });
            } else {
                hasAdjustedSize = true;
            }
        }

        @Override
        public void addView(View child, int index, LayoutParams params) {
            super.addView(child, index, params);
            if (hasAdjustedSize) {
                updateFirstChildView();
            }
        }

        @Override
        public void handleException(Throwable t) {
            getReactContext().handleException(new RuntimeException(t));
        }

        private ReactContext getReactContext() {
            return (ReactContext) getContext();
        }

        @Override
        public boolean onInterceptTouchEvent(MotionEvent event) {
            mJSTouchDispatcher.handleTouchEvent(event, getEventDispatcher());
            return super.onInterceptTouchEvent(event);
        }

        @Override
        public boolean onTouchEvent(MotionEvent event) {
            mJSTouchDispatcher.handleTouchEvent(event, getEventDispatcher());
            super.onTouchEvent(event);
            return true;
        }

        @Override
        public void onChildStartedNativeGesture(MotionEvent androidEvent) {
            mJSTouchDispatcher.onChildStartedNativeGesture(androidEvent, getEventDispatcher());
        }

        @Override
        public void requestDisallowInterceptTouchEvent(boolean disallowIntercept) {
        }

        private EventDispatcher getEventDispatcher() {
            ReactContext reactContext = getReactContext();
            return reactContext.getNativeModule(UIManagerModule.class).getEventDispatcher();
        }
    }

    public static class Crumb1 extends SceneActivity {} public static class Crumb2 extends SceneActivity {}
    public static class Crumb3 extends SceneActivity {} public static class Crumb4 extends SceneActivity {}
    public static class Crumb5 extends SceneActivity {} public static class Crumb6 extends SceneActivity {}
    public static class Crumb7 extends SceneActivity {} public static class Crumb8 extends SceneActivity {}
    public static class Crumb9 extends SceneActivity {} public static class Crumb10 extends SceneActivity {}
    public static class Crumb11 extends SceneActivity {} public static class Crumb12 extends SceneActivity {}
    public static class Crumb13 extends SceneActivity {} public static class Crumb14 extends SceneActivity {}
    public static class Crumb15 extends SceneActivity {} public static class Crumb16 extends SceneActivity {}
    public static class Crumb17 extends SceneActivity {} public static class Crumb18 extends SceneActivity {}
    public static class Crumb19 extends SceneActivity {} public static class Crumb20 extends SceneActivity {}

    public static Class getActivity(int crumb) {
        try {
            return Class.forName("com.navigation.reactnative.SceneActivity$Crumb" + crumb);
        } catch (ClassNotFoundException e) {
            return SceneActivity.class;
        }
    }

}
```

I also added the following lines to `settings.gradle` file:

```
include ':react-native-gesture-handler'
project(':react-native-gesture-handler').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-gesture-handler/android')

include ':app'
```

I get the following build error

```
> Task :navigation-react-native:compileDebugJavaWithJavac FAILED
/Users/arunmenon/Sites/reactnative/menuappnative/node_modules/navigation-react-native/android/app/src/main/java/com/navigation/reactnative/SceneActivity.java:24: error: package com.swmansion.gesturehandler.react does not exist
 import com.swmansion.gesturehandler.react.RNGestureHandlerEnabledRootView;
                                          ^
/Users/arunmenon/Sites/reactnative/menuappnative/node_modules/navigation-react-native/android/app/src/main/java/com/navigation/reactnative/SceneActivity.java:78: error: cannot find symbol
           return new RNGestureHandlerEnabledRootView(SceneActivity.this);
                      ^
  symbol: class RNGestureHandlerEnabledRootView
2 errors

FAILURE: Build failed with an exception.
```

I am guessing i will need to call/instantiate the equivalent of `new RNGestureHandlerPackage()` somewhere.

Im not sure how to proceed to completing the build for this approach to check if it works. My understanding is that this approach is akin to integrating `react-native-gesture-handler` into the `Navigation` project versus integrating into my project and extending where necessary. I am fine either way since I think the`Navigator` library for the native builds aims to provide for only native UI where available and doesn't actually have a wrapper that bridges functionality for UI elements for both the OSs as mentioned in [this comment](https://github.com/grahammendick/navigation/issues/250#issuecomment-504236309) where you say 

> The Navigation router will never support as many features as RNN because it has a much narrower scope. The Navigation router concentrates on navigation only so will never include a Modal or Drawer component. Also, the Navigation router only exposes the navigation features that are native to the platform. It doesn’t try to erase the platform differences. 

[My personal opinion though is that a `Drawer` component is part of the navigation solution since in almost every solution it resides to purely help navigate. And it is recognisable enough to warrant a second-thought about possibilities of integration in OSs that may not necessarily have the design pattern as standard.  Not very unlike a `NavigationBarIOS` for instance where there are provisions to navigate if necessary. A `Modal` though is more utility than a pure navigation-focussed UI pattern and anyway React does have a competent, if not robust, built-in modal solution ].

--- a/NavigationReactNative/src/android/app/src/main/java/com/navigation/reactnative/SceneView.java
+++ b/NavigationReactNative/src/android/app/src/main/java/com/navigation/reactnative/SceneView.java
@@ -1,5 +1,6 @@
 package com.navigation.reactnative;
 
+import android.view.View;
 import android.view.ViewGroup;
 
 import com.facebook.react.uimanager.ThemedReactContext;
@@ -9,6 +10,28 @@ public class SceneView extends ViewGroup {
     public SceneView(ThemedReactContext context) {
         super(context);
     }
+
+    @Override
+    protected void onAttachedToWindow() {
+        super.onAttachedToWindow();
+        View child = getChildAt(0);
+        if (child != null && child.getClass().getSimpleName().contains("DrawerLayout")) {
+            child.requestLayout();
+            post(measureAndLayout);
+        }
+    }
+
+    private final Runnable measureAndLayout =
+        new Runnable() {
+            @Override
+            public void run() {
+                View child = getChildAt(0);
+                child.measure(
+                    MeasureSpec.makeMeasureSpec(getWidth(), MeasureSpec.EXACTLY),
+                    MeasureSpec.makeMeasureSpec(getHeight(), MeasureSpec.EXACTLY));
+                child.layout(child.getLeft(), child.getTop(), child.getRight(), child.getBottom());
+            }
+        };
 
     @Override
     protected void onLayout(boolean changed, int l, int t, int r, int b) {

